File: algorithms/java/Graph/src/com/mahbub/algorithm/Tarjan_Demo.java
package com.mahbub.algorithm;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Stack;

public class Tarjan_Demo {

    static final char WHITE = 0;
    static final char GRAY = 1;
    static final char BLACK = 2;

    static int NODES, EDGES;
    static List<Integer>[] G;
    static char[] color;

    static int[] dfsNum, dfsLow;
    static int dfsNumberCounter;
    static Stack<Integer> S;
    static boolean[] inStack;

    static void tarjanSSC() {
        color = new char[NODES + 1];
        dfsNum = new int[NODES + 1];
        dfsLow = new int[NODES + 1];
        S = new Stack<>();
        inStack = new boolean[NODES + 1];

        for (int i = 0; i < NODES; i++) {
            if (color[i] == WHITE) strongConnect(i);
        }
    }

    static void strongConnect(int u) {
        dfsNum[u] = dfsLow[u] = dfsNumberCounter++;
        color[u] = GRAY;
        S.push(u);
        inStack[u] = true;

        for (int i = 0; i < G[u].size(); i++) {
            int v = G[u].get(i);
            if (color[v] == WHITE) {
                strongConnect(v);
                dfsLow[u] = Math.min(dfsLow[u], dfsLow[v]);
            } else if(inStack[v] == true) {
                dfsLow[u] = Math.min(dfsLow[u], dfsNum[v]);
            }
        }

        if (dfsNum[u] == dfsLow[u]) {
            int v = -1;
            while (v != u) {
                v = S.pop();
                inStack[v] = false;
                System.out.print(v + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Scanner scanner = inputFromFile();

        while (scanner.hasNext()) {
            NODES = scanner.nextInt();
            EDGES = scanner.nextInt();

            G = new List[NODES + 1];
            for (int i = 0; i <= NODES; i++) G[i] = new ArrayList<>();

            for (int i = 0; i < EDGES; i++) {
                int u = scanner.nextInt();
                int v = scanner.nextInt();

                G[u].add(v);
            }

            tarjanSSC();
            System.out.println();
        }
    }

    static Scanner inputFromFile() {
        try { return new Scanner(new FileInputStream("input.txt")); }
        catch (FileNotFoundException e) { e.printStackTrace(); }
        return null;
    }

    static Scanner inputFromSystem() {
        return new Scanner(System.in);
    }
}

/**

 INPUT:
 8 9
 0 1
 1 3
 3 2
 2 1
 3 4
 4 5
 5 7
 7 6
 6 4

 OUTPUT:
 6 7 5 4
 2 3 1
 0

 */

File: algorithms/kotlin/Graph/Tarjan_Demo.kt
package com.mahbub.algorithm.graph

import java.io.FileInputStream
import java.io.FileNotFoundException
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.E
import kotlin.math.min

class Tarjan_Demo {
    val WHITE : Byte = 0
    val GRAY : Byte = 1
    val BLACK : Byte = 2

    var NODES : Int = 0
    var EDGES : Int = 0
    var G : Array<MutableList<Int>> = arrayOf()
    var color : ByteArray = ByteArray(0)

    var dfsNum : IntArray = IntArray(0)
    var dfsLow : IntArray = IntArray(0)
    var dfsNumberCounter : Int = 0
    var S : Stack<Int> = Stack()
    var inStack : BooleanArray = BooleanArray(0)

    fun tarjanSSC() {
        color = ByteArray(NODES+1)
        dfsNum = IntArray(NODES+1)
        dfsLow = IntArray(NODES+1)
        S = Stack()
        inStack = BooleanArray(NODES+1)

        for (i in 0 until NODES)
            if (color[i] == WHITE) strongConnect(i)
    }

    fun strongConnect(u : Int) {
        dfsNumberCounter++
        dfsNum[u] = dfsNumberCounter
        dfsLow[u] = dfsNumberCounter
        color[u] = GRAY
        S.push(u)
        inStack[u] = true

        for (i in 0 until G[u].size) {
            var v = G[u].get(i)
            if (color[v] == WHITE) {
                strongConnect(v)
                dfsLow[u] = min(dfsLow[u], dfsLow[v])
            } else if (inStack[v] == true) {
                dfsLow[u] = min(dfsLow[u], dfsNum[v]);
            }
        }

        if (dfsNum[u] == dfsLow[u]) {
            var v = -1
            while (v != u) {
                v = S.pop()
                inStack[v] = false
                print("${v} ")
            }
            println()
        }
    }

    fun fakeMain() {
        var scanner = inputFromFile()

        NODES = scanner!!.nextInt()
        EDGES = scanner!!.nextInt()

        G = Array(NODES+1) { ArrayList<Int>() }

        for (i in 0 until EDGES) {
            var u = scanner!!.nextInt()
            var v = scanner!!.nextInt()

            G[u].add(v)
        }

        tarjanSSC()
        println()
    }
}

fun main(args : Array<String>) {
    Tarjan_Demo().fakeMain()
}

private fun inputFromFile() : Scanner? {
    return try { Scanner(FileInputStream("input.txt")) }
    catch (e : FileNotFoundException) { null }
}

private fun inputFromSystem() : Scanner? {
    return Scanner(System.`in`)
}